src/uu/sed/src/delimited_parser.rs

```diff
@@ -0,0 +1,918 @@
+// Parse delimited character sequences
+//
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2025 Diomidis Spinellis
+//
+// This file is part of the uutils sed package.
+// It is licensed under the MIT License.
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+use crate::script_char_provider::ScriptCharProvider;
+use crate::script_line_provider::ScriptLineProvider;
+use std::char;
+use uucore::error::{UResult, USimpleError};
+
+// Fail with msg as a compile error at the current location
+pub fn compilation_error<T>(
+    lines: &ScriptLineProvider,
+    line: &ScriptCharProvider,
+    msg: impl ToString,
+) -> UResult<T> {
+    Err(USimpleError::new(
+        1,
+        format!(
+            "{}:{}:{}: error: {}",
+            lines.get_input_name(),
+            lines.get_line_number(),
+            line.get_pos(),
+            msg.to_string()
+        ),
+    ))
+}
+
+/// Return true if c is a valid octal digit
+fn is_ascii_octal_digit(c: char) -> bool {
+    matches!(c, '0'..='7')
+}
+
+/// Parse a numeric character escape and return the corresponding char.
+/// Advance line to the first character not part of the escape.
+/// ndigits is the number of allowed digits and radix is the value's
+/// radix (e.g. 8, 10, 16 for octal, decimal, and hex escapes).
+/// For values up to 3 ndigits is the maximum number of allowed digits,
+/// for values above 3 ndigits is the exact number of allowed digits.
+/// Return `None` if no valid character has been specified.
+fn parse_numeric_escape(
+    line: &mut ScriptCharProvider,
+    is_allowed_char: fn(char) -> bool,
+    ndigits: usize,
+    radix: u32,
+) -> Option<char> {
+    let mut valid_chars = Vec::new();
+
+    for _ in 0..ndigits {
+        if !line.eol() && is_allowed_char(line.current()) {
+            valid_chars.push(line.current());
+            line.advance();
+        } else {
+            break;
+        }
+    }
+
+    if valid_chars.is_empty() {
+        return None;
+    }
+
+    if ndigits > 3 && valid_chars.len() != ndigits {
+        line.retreat(valid_chars.len());
+        return None;
+    }
+
+    let char_string: String = valid_chars.into_iter().collect();
+    match u32::from_str_radix(&char_string, radix)
+        .ok()
+        .and_then(char::from_u32)
+    {
+        Some(decoded) => Some(decoded),
+        None => panic!("Unable to decode numeric character escape."),
```

Comment:
same comment about the panic :)

Reply:
Same here: preceding lines have already verified that this is a valid sequence for the specified radix. The panic should only be triggered if there's a code error not a user error.